{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4c6a698f",
   "metadata": {},
   "source": [
    "# NMF lab"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df3f3f0f",
   "metadata": {},
   "source": [
    "### Task (45 minutes) \n",
    "\n",
    "\n",
    "__Important__\n",
    "- This task is optimized for speedrun and not quality. Normaly you would look into more components-\n",
    "\n",
    "\n",
    "Data: MOFA's CLL dataset\n",
    "\n",
    "1. Load and prepare data, then perform NMR on the joint representation of the methylation and mRNA datasets.\n",
    "2. Assume that there are two cancer subtypes and cluster them :)\n",
    "3. Find the driving features and verify their functionality.\n",
    "\n",
    "\n",
    "Code:\n",
    "- Lab example uses a NMF implementation in Python:\n",
    "    - https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.NMF.html#sklearn.decomposition.NMF\n",
    "- For R, feel free to use this package:\n",
    "    - https://cran.r-project.org/web/packages/NMF/index.html"
   ]
  },
  {
   "cell_type": "raw",
   "id": "6089f54a",
   "metadata": {},
   "source": [
    "conda create -n iocourse\n",
    "conda activate iocourse\n",
    "conda install jupyterlab pandas numpy scikit-learn\n",
    "pip install snfpy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e773cf45-6218-4dc2-84d5-6e85fef8ce45",
   "metadata": {},
   "source": [
    "Load the datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0ce2d61d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_loc = \"C:\\\\mydata\\\\work\\\\io_course\\\\data\\\\\"\n",
    "\n",
    "import pandas as pd\n",
    "df_meth = pd.read_csv(data_loc + \"CLL_data_Methylation.csv\", index_col=0)\n",
    "df_mrna = pd.read_csv(data_loc + \"CLL_data_mRNA.csv\", index_col=0)\n",
    "\n",
    "# drop nans by column\n",
    "df_mrna = df_mrna.dropna(axis='columns')\n",
    "df_meth = df_meth.dropna(axis='columns')\n",
    "\n",
    "df_mrna = df_mrna.T\n",
    "df_meth = df_meth.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "13f26555",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cg10146935</th>\n",
       "      <th>cg26837773</th>\n",
       "      <th>cg17801765</th>\n",
       "      <th>cg13244315</th>\n",
       "      <th>cg06181703</th>\n",
       "      <th>cg19626656</th>\n",
       "      <th>cg15207968</th>\n",
       "      <th>cg12755103</th>\n",
       "      <th>cg23651812</th>\n",
       "      <th>cg14287724</th>\n",
       "      <th>...</th>\n",
       "      <th>cg07016730</th>\n",
       "      <th>cg25152348</th>\n",
       "      <th>cg08425796</th>\n",
       "      <th>cg05418105</th>\n",
       "      <th>cg22249529</th>\n",
       "      <th>cg07600533</th>\n",
       "      <th>cg08260245</th>\n",
       "      <th>cg19112186</th>\n",
       "      <th>cg10770023</th>\n",
       "      <th>cg00270625</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>H045</th>\n",
       "      <td>1.811086</td>\n",
       "      <td>-5.172572</td>\n",
       "      <td>5.411526</td>\n",
       "      <td>-0.118825</td>\n",
       "      <td>5.120384</td>\n",
       "      <td>0.145951</td>\n",
       "      <td>-3.436869</td>\n",
       "      <td>-3.844246</td>\n",
       "      <td>2.075422</td>\n",
       "      <td>3.501829</td>\n",
       "      <td>...</td>\n",
       "      <td>3.547843</td>\n",
       "      <td>0.060132</td>\n",
       "      <td>4.442026</td>\n",
       "      <td>2.861301</td>\n",
       "      <td>5.246799</td>\n",
       "      <td>3.901933</td>\n",
       "      <td>5.713831</td>\n",
       "      <td>5.703520</td>\n",
       "      <td>5.166255</td>\n",
       "      <td>4.911655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H109</th>\n",
       "      <td>-3.997508</td>\n",
       "      <td>1.594870</td>\n",
       "      <td>5.412693</td>\n",
       "      <td>1.043871</td>\n",
       "      <td>1.279480</td>\n",
       "      <td>-3.928433</td>\n",
       "      <td>2.989245</td>\n",
       "      <td>0.393004</td>\n",
       "      <td>4.800121</td>\n",
       "      <td>3.159201</td>\n",
       "      <td>...</td>\n",
       "      <td>0.887926</td>\n",
       "      <td>-0.214753</td>\n",
       "      <td>4.561187</td>\n",
       "      <td>3.919911</td>\n",
       "      <td>5.058302</td>\n",
       "      <td>2.634941</td>\n",
       "      <td>5.107460</td>\n",
       "      <td>1.326244</td>\n",
       "      <td>0.677912</td>\n",
       "      <td>5.281115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H024</th>\n",
       "      <td>-2.844313</td>\n",
       "      <td>0.161170</td>\n",
       "      <td>0.365706</td>\n",
       "      <td>-4.219236</td>\n",
       "      <td>0.721100</td>\n",
       "      <td>-3.418859</td>\n",
       "      <td>-3.250385</td>\n",
       "      <td>-2.691305</td>\n",
       "      <td>0.534854</td>\n",
       "      <td>-4.629484</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.486709</td>\n",
       "      <td>0.121749</td>\n",
       "      <td>-2.841373</td>\n",
       "      <td>-3.607177</td>\n",
       "      <td>0.765651</td>\n",
       "      <td>1.516759</td>\n",
       "      <td>5.676245</td>\n",
       "      <td>5.488636</td>\n",
       "      <td>4.221828</td>\n",
       "      <td>5.379716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H056</th>\n",
       "      <td>-3.338656</td>\n",
       "      <td>-2.093433</td>\n",
       "      <td>0.373634</td>\n",
       "      <td>-1.592196</td>\n",
       "      <td>4.047059</td>\n",
       "      <td>0.226601</td>\n",
       "      <td>2.377386</td>\n",
       "      <td>-2.775075</td>\n",
       "      <td>0.419985</td>\n",
       "      <td>0.312388</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.238214</td>\n",
       "      <td>0.137862</td>\n",
       "      <td>-3.964855</td>\n",
       "      <td>-2.270940</td>\n",
       "      <td>-2.631909</td>\n",
       "      <td>-3.884756</td>\n",
       "      <td>5.950338</td>\n",
       "      <td>5.354059</td>\n",
       "      <td>4.934536</td>\n",
       "      <td>5.366823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H079</th>\n",
       "      <td>-0.019362</td>\n",
       "      <td>3.748980</td>\n",
       "      <td>5.412010</td>\n",
       "      <td>1.416418</td>\n",
       "      <td>5.237422</td>\n",
       "      <td>0.324213</td>\n",
       "      <td>-0.647632</td>\n",
       "      <td>-3.098837</td>\n",
       "      <td>5.397188</td>\n",
       "      <td>3.410770</td>\n",
       "      <td>...</td>\n",
       "      <td>2.758021</td>\n",
       "      <td>0.021011</td>\n",
       "      <td>0.673296</td>\n",
       "      <td>3.455230</td>\n",
       "      <td>-3.140733</td>\n",
       "      <td>-4.238106</td>\n",
       "      <td>6.040756</td>\n",
       "      <td>5.584746</td>\n",
       "      <td>5.095111</td>\n",
       "      <td>5.338470</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 4248 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      cg10146935  cg26837773  cg17801765  cg13244315  cg06181703  cg19626656  \\\n",
       "H045    1.811086   -5.172572    5.411526   -0.118825    5.120384    0.145951   \n",
       "H109   -3.997508    1.594870    5.412693    1.043871    1.279480   -3.928433   \n",
       "H024   -2.844313    0.161170    0.365706   -4.219236    0.721100   -3.418859   \n",
       "H056   -3.338656   -2.093433    0.373634   -1.592196    4.047059    0.226601   \n",
       "H079   -0.019362    3.748980    5.412010    1.416418    5.237422    0.324213   \n",
       "\n",
       "      cg15207968  cg12755103  cg23651812  cg14287724  ...  cg07016730  \\\n",
       "H045   -3.436869   -3.844246    2.075422    3.501829  ...    3.547843   \n",
       "H109    2.989245    0.393004    4.800121    3.159201  ...    0.887926   \n",
       "H024   -3.250385   -2.691305    0.534854   -4.629484  ...   -4.486709   \n",
       "H056    2.377386   -2.775075    0.419985    0.312388  ...   -4.238214   \n",
       "H079   -0.647632   -3.098837    5.397188    3.410770  ...    2.758021   \n",
       "\n",
       "      cg25152348  cg08425796  cg05418105  cg22249529  cg07600533  cg08260245  \\\n",
       "H045    0.060132    4.442026    2.861301    5.246799    3.901933    5.713831   \n",
       "H109   -0.214753    4.561187    3.919911    5.058302    2.634941    5.107460   \n",
       "H024    0.121749   -2.841373   -3.607177    0.765651    1.516759    5.676245   \n",
       "H056    0.137862   -3.964855   -2.270940   -2.631909   -3.884756    5.950338   \n",
       "H079    0.021011    0.673296    3.455230   -3.140733   -4.238106    6.040756   \n",
       "\n",
       "      cg19112186  cg10770023  cg00270625  \n",
       "H045    5.703520    5.166255    4.911655  \n",
       "H109    1.326244    0.677912    5.281115  \n",
       "H024    5.488636    4.221828    5.379716  \n",
       "H056    5.354059    4.934536    5.366823  \n",
       "H079    5.584746    5.095111    5.338470  \n",
       "\n",
       "[5 rows x 4248 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meth.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0021f9d7-b93c-4733-9128-c528f75fd5c0",
   "metadata": {},
   "source": [
    "To ensure non-negativity, add the negative modality data as new features. Just some ugly code, not important. :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "60866a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = df_meth.columns.copy()\n",
    "columns = {}\n",
    "for c in cols:\n",
    "    mask = df_meth[c] < 0\n",
    "    columns[c + '_p'] = df_meth[c].mask(mask)\n",
    "    columns[c + '_n'] = - df_meth[c].mask(~mask)\n",
    "df_meth = pd.concat(list(columns.values()), keys=list(columns.keys()), axis=1)  \n",
    "df_meth = df_meth.fillna(0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "034712a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H045</th>\n",
       "      <th>H109</th>\n",
       "      <th>H024</th>\n",
       "      <th>H056</th>\n",
       "      <th>H079</th>\n",
       "      <th>H164</th>\n",
       "      <th>H059</th>\n",
       "      <th>H167</th>\n",
       "      <th>H113</th>\n",
       "      <th>H049</th>\n",
       "      <th>...</th>\n",
       "      <th>H106</th>\n",
       "      <th>H176</th>\n",
       "      <th>H136</th>\n",
       "      <th>H178</th>\n",
       "      <th>H166</th>\n",
       "      <th>H174</th>\n",
       "      <th>H177</th>\n",
       "      <th>H259</th>\n",
       "      <th>H175</th>\n",
       "      <th>H179</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cg10146935_p</th>\n",
       "      <td>1.811086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg10146935_n</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.997508</td>\n",
       "      <td>2.844313</td>\n",
       "      <td>3.338656</td>\n",
       "      <td>0.019362</td>\n",
       "      <td>2.485997</td>\n",
       "      <td>1.460211</td>\n",
       "      <td>4.952291</td>\n",
       "      <td>2.980209</td>\n",
       "      <td>0.091713</td>\n",
       "      <td>...</td>\n",
       "      <td>4.946979</td>\n",
       "      <td>0.838754</td>\n",
       "      <td>4.684879</td>\n",
       "      <td>5.077259</td>\n",
       "      <td>0.625954</td>\n",
       "      <td>4.918812</td>\n",
       "      <td>4.727493</td>\n",
       "      <td>5.193812</td>\n",
       "      <td>4.437189</td>\n",
       "      <td>5.060459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg26837773_p</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.594870</td>\n",
       "      <td>0.161170</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.748980</td>\n",
       "      <td>0.060530</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.547577</td>\n",
       "      <td>2.440098</td>\n",
       "      <td>2.940767</td>\n",
       "      <td>...</td>\n",
       "      <td>0.199544</td>\n",
       "      <td>0.341156</td>\n",
       "      <td>3.496116</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.214163</td>\n",
       "      <td>2.036858</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.043775</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg26837773_n</th>\n",
       "      <td>5.172572</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.093433</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.472232</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.821910</td>\n",
       "      <td>0.858200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.816088</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.345652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cg17801765_p</th>\n",
       "      <td>5.411526</td>\n",
       "      <td>5.412693</td>\n",
       "      <td>0.365706</td>\n",
       "      <td>0.373634</td>\n",
       "      <td>5.412010</td>\n",
       "      <td>5.268908</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.337081</td>\n",
       "      <td>0.749546</td>\n",
       "      <td>0.426493</td>\n",
       "      <td>...</td>\n",
       "      <td>0.574758</td>\n",
       "      <td>0.707883</td>\n",
       "      <td>0.438992</td>\n",
       "      <td>4.873615</td>\n",
       "      <td>0.753594</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.547390</td>\n",
       "      <td>4.086683</td>\n",
       "      <td>0.135581</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 196 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  H045      H109      H024      H056      H079      H164  \\\n",
       "cg10146935_p  1.811086  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "cg10146935_n  0.000000  3.997508  2.844313  3.338656  0.019362  2.485997   \n",
       "cg26837773_p  0.000000  1.594870  0.161170  0.000000  3.748980  0.060530   \n",
       "cg26837773_n  5.172572  0.000000  0.000000  2.093433  0.000000  0.000000   \n",
       "cg17801765_p  5.411526  5.412693  0.365706  0.373634  5.412010  5.268908   \n",
       "\n",
       "                  H059      H167      H113      H049  ...      H106      H176  \\\n",
       "cg10146935_p  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "cg10146935_n  1.460211  4.952291  2.980209  0.091713  ...  4.946979  0.838754   \n",
       "cg26837773_p  0.000000  0.547577  2.440098  2.940767  ...  0.199544  0.341156   \n",
       "cg26837773_n  3.472232  0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "cg17801765_p  0.000000  5.337081  0.749546  0.426493  ...  0.574758  0.707883   \n",
       "\n",
       "                  H136      H178      H166      H174      H177      H259  \\\n",
       "cg10146935_p  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "cg10146935_n  4.684879  5.077259  0.625954  4.918812  4.727493  5.193812   \n",
       "cg26837773_p  3.496116  0.000000  0.000000  3.214163  2.036858  0.000000   \n",
       "cg26837773_n  0.000000  4.821910  0.858200  0.000000  0.000000  0.816088   \n",
       "cg17801765_p  0.438992  4.873615  0.753594  0.000000  0.000000  0.547390   \n",
       "\n",
       "                  H175      H179  \n",
       "cg10146935_p  0.000000  0.000000  \n",
       "cg10146935_n  4.437189  5.060459  \n",
       "cg26837773_p  4.043775  0.000000  \n",
       "cg26837773_n  0.000000  2.345652  \n",
       "cg17801765_p  4.086683  0.135581  \n",
       "\n",
       "[5 rows x 196 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meth.T.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2dd6958-b45c-4e13-9295-b4d77aecb0d3",
   "metadata": {},
   "source": [
    "Apply the frobenius norms to normalize the datasets, then concatenate the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "7ac9a05d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_mrna.T\n",
    "df = df/df.mean()\n",
    "fro = df.apply(lambda x: (x**2).sum()**.5, axis=0)\n",
    "df_mrna = df / fro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "405a4dd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_meth.T\n",
    "df = df/df.mean()\n",
    "fro = df.apply(lambda x: (x**2).sum()**.5, axis=0)\n",
    "df_meth = df / fro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "306a9d4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13496, 135)\n"
     ]
    }
   ],
   "source": [
    "X = pd.concat([df_mrna, df_meth])\n",
    "X = X.dropna(axis='columns')\n",
    "print(X.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "535a45be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H045</th>\n",
       "      <th>H109</th>\n",
       "      <th>H024</th>\n",
       "      <th>H056</th>\n",
       "      <th>H079</th>\n",
       "      <th>H164</th>\n",
       "      <th>H059</th>\n",
       "      <th>H167</th>\n",
       "      <th>H113</th>\n",
       "      <th>H049</th>\n",
       "      <th>...</th>\n",
       "      <th>H271</th>\n",
       "      <th>H006</th>\n",
       "      <th>H084</th>\n",
       "      <th>H260</th>\n",
       "      <th>H192</th>\n",
       "      <th>H070</th>\n",
       "      <th>H255</th>\n",
       "      <th>H135</th>\n",
       "      <th>H247</th>\n",
       "      <th>H066</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000244734</th>\n",
       "      <td>0.009022</td>\n",
       "      <td>0.005391</td>\n",
       "      <td>0.020382</td>\n",
       "      <td>0.026755</td>\n",
       "      <td>0.012140</td>\n",
       "      <td>0.005026</td>\n",
       "      <td>0.010195</td>\n",
       "      <td>0.003092</td>\n",
       "      <td>0.004567</td>\n",
       "      <td>0.005180</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008746</td>\n",
       "      <td>0.017387</td>\n",
       "      <td>0.021283</td>\n",
       "      <td>0.024401</td>\n",
       "      <td>0.013766</td>\n",
       "      <td>0.006895</td>\n",
       "      <td>0.006624</td>\n",
       "      <td>0.003180</td>\n",
       "      <td>0.017701</td>\n",
       "      <td>0.007995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000158528</th>\n",
       "      <td>0.023238</td>\n",
       "      <td>0.026321</td>\n",
       "      <td>0.004801</td>\n",
       "      <td>0.006514</td>\n",
       "      <td>0.023816</td>\n",
       "      <td>0.022488</td>\n",
       "      <td>0.011320</td>\n",
       "      <td>0.005297</td>\n",
       "      <td>0.005619</td>\n",
       "      <td>0.005180</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007797</td>\n",
       "      <td>0.007975</td>\n",
       "      <td>0.005282</td>\n",
       "      <td>0.025212</td>\n",
       "      <td>0.006281</td>\n",
       "      <td>0.003239</td>\n",
       "      <td>0.025178</td>\n",
       "      <td>0.003180</td>\n",
       "      <td>0.006260</td>\n",
       "      <td>0.024524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198478</th>\n",
       "      <td>0.017657</td>\n",
       "      <td>0.005391</td>\n",
       "      <td>0.025392</td>\n",
       "      <td>0.016334</td>\n",
       "      <td>0.009752</td>\n",
       "      <td>0.024930</td>\n",
       "      <td>0.007419</td>\n",
       "      <td>0.008332</td>\n",
       "      <td>0.010429</td>\n",
       "      <td>0.005966</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004636</td>\n",
       "      <td>0.010719</td>\n",
       "      <td>0.018570</td>\n",
       "      <td>0.020788</td>\n",
       "      <td>0.012456</td>\n",
       "      <td>0.003239</td>\n",
       "      <td>0.005890</td>\n",
       "      <td>0.003180</td>\n",
       "      <td>0.016220</td>\n",
       "      <td>0.023507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175445</th>\n",
       "      <td>0.025108</td>\n",
       "      <td>0.021643</td>\n",
       "      <td>0.003135</td>\n",
       "      <td>0.003081</td>\n",
       "      <td>0.026607</td>\n",
       "      <td>0.021274</td>\n",
       "      <td>0.023107</td>\n",
       "      <td>0.023463</td>\n",
       "      <td>0.004567</td>\n",
       "      <td>0.005966</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004636</td>\n",
       "      <td>0.018250</td>\n",
       "      <td>0.018723</td>\n",
       "      <td>0.023259</td>\n",
       "      <td>0.003126</td>\n",
       "      <td>0.009398</td>\n",
       "      <td>0.023147</td>\n",
       "      <td>0.003180</td>\n",
       "      <td>0.011393</td>\n",
       "      <td>0.021184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000174469</th>\n",
       "      <td>0.005235</td>\n",
       "      <td>0.025055</td>\n",
       "      <td>0.003135</td>\n",
       "      <td>0.027334</td>\n",
       "      <td>0.010923</td>\n",
       "      <td>0.021449</td>\n",
       "      <td>0.016399</td>\n",
       "      <td>0.005297</td>\n",
       "      <td>0.004567</td>\n",
       "      <td>0.026762</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028580</td>\n",
       "      <td>0.020996</td>\n",
       "      <td>0.024090</td>\n",
       "      <td>0.018633</td>\n",
       "      <td>0.010119</td>\n",
       "      <td>0.026308</td>\n",
       "      <td>0.003097</td>\n",
       "      <td>0.026659</td>\n",
       "      <td>0.019995</td>\n",
       "      <td>0.021093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     H045      H109      H024      H056      H079      H164  \\\n",
       "ENSG00000244734  0.009022  0.005391  0.020382  0.026755  0.012140  0.005026   \n",
       "ENSG00000158528  0.023238  0.026321  0.004801  0.006514  0.023816  0.022488   \n",
       "ENSG00000198478  0.017657  0.005391  0.025392  0.016334  0.009752  0.024930   \n",
       "ENSG00000175445  0.025108  0.021643  0.003135  0.003081  0.026607  0.021274   \n",
       "ENSG00000174469  0.005235  0.025055  0.003135  0.027334  0.010923  0.021449   \n",
       "\n",
       "                     H059      H167      H113      H049  ...      H271  \\\n",
       "ENSG00000244734  0.010195  0.003092  0.004567  0.005180  ...  0.008746   \n",
       "ENSG00000158528  0.011320  0.005297  0.005619  0.005180  ...  0.007797   \n",
       "ENSG00000198478  0.007419  0.008332  0.010429  0.005966  ...  0.004636   \n",
       "ENSG00000175445  0.023107  0.023463  0.004567  0.005966  ...  0.004636   \n",
       "ENSG00000174469  0.016399  0.005297  0.004567  0.026762  ...  0.028580   \n",
       "\n",
       "                     H006      H084      H260      H192      H070      H255  \\\n",
       "ENSG00000244734  0.017387  0.021283  0.024401  0.013766  0.006895  0.006624   \n",
       "ENSG00000158528  0.007975  0.005282  0.025212  0.006281  0.003239  0.025178   \n",
       "ENSG00000198478  0.010719  0.018570  0.020788  0.012456  0.003239  0.005890   \n",
       "ENSG00000175445  0.018250  0.018723  0.023259  0.003126  0.009398  0.023147   \n",
       "ENSG00000174469  0.020996  0.024090  0.018633  0.010119  0.026308  0.003097   \n",
       "\n",
       "                     H135      H247      H066  \n",
       "ENSG00000244734  0.003180  0.017701  0.007995  \n",
       "ENSG00000158528  0.003180  0.006260  0.024524  \n",
       "ENSG00000198478  0.003180  0.016220  0.023507  \n",
       "ENSG00000175445  0.003180  0.011393  0.021184  \n",
       "ENSG00000174469  0.026659  0.019995  0.021093  \n",
       "\n",
       "[5 rows x 135 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f90c782d-4ad1-4721-a0e5-56457053a3b7",
   "metadata": {},
   "source": [
    "### Fiting an NMF model\n",
    "\n",
    "- So now we fit a two component NMF model and check results.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9ee1cf59",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import NMF\n",
    "model = NMF(n_components=2, init='random', random_state=0)\n",
    "W = model.fit_transform(X)\n",
    "H = model.components_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ff81a127",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, 135) (13496, 2)\n"
     ]
    }
   ],
   "source": [
    "print(H.shape, W.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "a570828a-8e89-465c-a94d-ae503bef8008",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.12609143, 0.0962925 ],\n",
       "       [0.04392555, 0.11771427],\n",
       "       [0.09126849, 0.10719918],\n",
       "       ...,\n",
       "       [0.00193168, 0.01028896],\n",
       "       [0.17476978, 0.07858805],\n",
       "       [0.00302261, 0.02073489]])"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "W"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "60285549",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>latent_1</th>\n",
       "      <th>latent_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000244734</th>\n",
       "      <td>0.126091</td>\n",
       "      <td>0.096292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000158528</th>\n",
       "      <td>0.043926</td>\n",
       "      <td>0.117714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198478</th>\n",
       "      <td>0.091268</td>\n",
       "      <td>0.107199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000175445</th>\n",
       "      <td>0.065235</td>\n",
       "      <td>0.146637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000174469</th>\n",
       "      <td>0.197567</td>\n",
       "      <td>0.098666</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 latent_1  latent_2\n",
       "ENSG00000244734  0.126091  0.096292\n",
       "ENSG00000158528  0.043926  0.117714\n",
       "ENSG00000198478  0.091268  0.107199\n",
       "ENSG00000175445  0.065235  0.146637\n",
       "ENSG00000174469  0.197567  0.098666"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#W_df = pd.DataFrame(W, columns=X.index.values)\n",
    "W_df = pd.DataFrame(W, columns=[\"latent_1\", \"latent_2\"], index = X.index)\n",
    "W_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1d2306c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H045</th>\n",
       "      <th>H109</th>\n",
       "      <th>H024</th>\n",
       "      <th>H056</th>\n",
       "      <th>H079</th>\n",
       "      <th>H164</th>\n",
       "      <th>H059</th>\n",
       "      <th>H167</th>\n",
       "      <th>H113</th>\n",
       "      <th>H049</th>\n",
       "      <th>...</th>\n",
       "      <th>H271</th>\n",
       "      <th>H006</th>\n",
       "      <th>H084</th>\n",
       "      <th>H260</th>\n",
       "      <th>H192</th>\n",
       "      <th>H070</th>\n",
       "      <th>H255</th>\n",
       "      <th>H135</th>\n",
       "      <th>H247</th>\n",
       "      <th>H066</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>latent_1</th>\n",
       "      <td>0.005286</td>\n",
       "      <td>0.001123</td>\n",
       "      <td>0.116696</td>\n",
       "      <td>0.1156</td>\n",
       "      <td>0.008202</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.108295</td>\n",
       "      <td>0.004208</td>\n",
       "      <td>0.068335</td>\n",
       "      <td>0.053253</td>\n",
       "      <td>...</td>\n",
       "      <td>0.112052</td>\n",
       "      <td>0.117386</td>\n",
       "      <td>0.000702</td>\n",
       "      <td>0.005236</td>\n",
       "      <td>0.05371</td>\n",
       "      <td>0.112368</td>\n",
       "      <td>0.012251</td>\n",
       "      <td>0.113764</td>\n",
       "      <td>0.116019</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>latent_2</th>\n",
       "      <td>0.144156</td>\n",
       "      <td>0.149550</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.140064</td>\n",
       "      <td>0.150715</td>\n",
       "      <td>0.009437</td>\n",
       "      <td>0.145258</td>\n",
       "      <td>0.061816</td>\n",
       "      <td>0.082893</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004414</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.150970</td>\n",
       "      <td>0.142550</td>\n",
       "      <td>0.08071</td>\n",
       "      <td>0.004421</td>\n",
       "      <td>0.134242</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.152084</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              H045      H109      H024    H056      H079      H164      H059  \\\n",
       "latent_1  0.005286  0.001123  0.116696  0.1156  0.008202  0.000000  0.108295   \n",
       "latent_2  0.144156  0.149550  0.000000  0.0000  0.140064  0.150715  0.009437   \n",
       "\n",
       "              H167      H113      H049  ...      H271      H006      H084  \\\n",
       "latent_1  0.004208  0.068335  0.053253  ...  0.112052  0.117386  0.000702   \n",
       "latent_2  0.145258  0.061816  0.082893  ...  0.004414  0.000000  0.150970   \n",
       "\n",
       "              H260     H192      H070      H255      H135      H247      H066  \n",
       "latent_1  0.005236  0.05371  0.112368  0.012251  0.113764  0.116019  0.000000  \n",
       "latent_2  0.142550  0.08071  0.004421  0.134242  0.000000  0.000000  0.152084  \n",
       "\n",
       "[2 rows x 135 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "H_df = pd.DataFrame(H, columns=X.columns, index = [\"latent_1\", \"latent_2\"])\n",
    "H_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7df0bf27-18e3-47b7-925f-f073d6fc2a40",
   "metadata": {},
   "source": [
    "__Clustering__\n",
    "\n",
    "- Clustering the samples on two subtypes doesn't require any specific clustering algorithms :) But for more subtypes you are welcome to measure up against MOFA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5df3fb9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1,\n",
       "       0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0,\n",
       "       0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0,\n",
       "       0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0,\n",
       "       0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0,\n",
       "       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1,\n",
       "       0, 0, 1], dtype=int64)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "clusters = np.argmax(H, axis = 0)\n",
    "clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21c8243e-5719-42b9-863a-54556d3085ce",
   "metadata": {},
   "source": [
    "### NMF analysis\n",
    "\n",
    "What samples drive our first component (cluster)?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "03a6f990-f7f1-4dde-bd8f-971f0ebf8b9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "H022    0.117501\n",
       "H006    0.117386\n",
       "H110    0.116995\n",
       "H236    0.116745\n",
       "H024    0.116696\n",
       "          ...   \n",
       "H042    0.000000\n",
       "H077    0.000000\n",
       "H008    0.000000\n",
       "H027    0.000000\n",
       "H066    0.000000\n",
       "Name: latent_1, Length: 135, dtype: float64"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = H_df.loc[\"latent_1\",].sort_values(ascending=False)\n",
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "a6501a9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['H022', 'H006', 'H110', 'H236', 'H024', 'H099', 'H038', 'H247',\n",
       "       'H133', 'H060'], dtype=object)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top10 = samples.index.values[0:10]\n",
    "top10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f9a02e9-41c1-472b-a72f-466a47aaabcd",
   "metadata": {},
   "source": [
    "Similarly, what transcripts ID and methylation probes are driving the signal in the same component?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "74e85691-f103-4831-a779-dafc404af8be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENSG00000092820    0.298223\n",
       "ENSG00000067082    0.293701\n",
       "ENSG00000142102    0.285527\n",
       "ENSG00000177606    0.282739\n",
       "ENSG00000156738    0.281571\n",
       "                     ...   \n",
       "cg18766900_p       0.000000\n",
       "cg06528737_p       0.000000\n",
       "cg10853431_p       0.000000\n",
       "cg07240413_p       0.000000\n",
       "cg04387658_p       0.000000\n",
       "Name: latent_1, Length: 13496, dtype: float64"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first component of W\n",
    "features = W_df.iloc[:,0].sort_values(ascending=False)\n",
    "features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55663889-761d-4059-814d-d7ea9ea3aecd",
   "metadata": {},
   "source": [
    "Let's get the first ten genes and the first ten probes..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "e6d5c723-b0d4-4253-a8a1-9de0932e9a74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000092820', 'ENSG00000067082', 'ENSG00000142102', 'ENSG00000177606', 'ENSG00000156738', 'ENSG00000077238', 'ENSG00000215301', 'ENSG00000087074', 'ENSG00000112149', 'ENSG00000143297']\n",
      "['cg04813697_n', 'cg01360627_n', 'cg06401414_n', 'cg03633073_p', 'cg04694619_n', 'cg02650512_n', 'cg05277504_p', 'cg04173586_n', 'cg05656688_n', 'cg10379077_n']\n"
     ]
    }
   ],
   "source": [
    "top_g = [f for f in features.index.values if f[:2]==\"EN\"]\n",
    "top_p = [f for f in features.index.values if f[:2]==\"cg\"]\n",
    "print(top_g[:10])\n",
    "print(top_p[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c03dcdc4-b1aa-41a3-8f16-099b39474464",
   "metadata": {},
   "source": [
    "### Does is look simpler than MOFA? \n",
    "\n",
    "well...\n",
    "\n",
    "Now, for finer points, expand this study:\n",
    "- Use a range of K values and fit the model then cluster via Kmeans and perform silhouette coefficient testing.\n",
    "    - https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html\n",
    "    - https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html\n",
    "- Observe convergence by computing the mean squared error.\n",
    "    - https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html\n",
    "- Use the other omics tables as well, for .. something.\n",
    "- Compare with the results from the MOFA analysis\n",
    "https://bioconductor.riken.jp/packages/3.9/bioc/vignettes/MOFA/inst/doc/MOFA_example_CLL.html\n",
    "\n",
    "Shameless self promotion:\n",
    "- I hold a course of data science using advanced python, that covers everything, from numerical computing to deep learning. For details check the SciLifeLab page."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1202f50-f230-44e8-8aab-531630570136",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
